Add primitive type tutorial blog (#321)
JuliaLang/www.julialang.org: blog/_posts/2019-03-02-custom-primitiv-type.md

---
layout: post
title: Declare primitive type in Julia
author: 鄭景文（Peter Cheng）, 杜岳華（Yueh-Hua Tu）
---

> Julia provide some APIs for you to declare your own primitive types in Julia. However, when you open [the official docs about primitive type](https://docs.julialang.org/en/v1/manual/types/#Primitive-Types-1), you still don't know how to make your new primitive types work in your code. Therefore, here is a simple tutorial that tell you how to do so.

Reviewer: ```julia
Julia provides
```

No need to repeat "in Julia" at the end :)

## Introduction

If you want define a primitve type, first, you have to declare it with the name and the number of bits it take with the `primitive type` keyword. 

    primitive type «name» «bits» end
    primitive type «name» <: «supertype» «bits» end

> Currently, Julia only support sizes that are multiples of 8 bits. So like the `Bool` only take 1 bit, you still need to declare it as 8 bits. For example:

Reviewer: `Julia only supports`
So even though `Bool` only takes 1 bit, you need to declare it as 8 bits

```julia
    primitive type Bool <: Integer 8 end
```

Up to here, we have almost everything from the docs. Next, we are going to show how 

> to use our own primitive type with a simple example.

Reviewer: ```suggestion
to define and use our own primitive type with a simple example.
```

## The problem

After declaring our new primitive type, that's say we do the following:

```julia
    primitive type A 72 end
```

> But soon you will find that you don't know how to create a instance of `A` because Julia don't have a default constructor for a new primitive type. And if you think further, you will notice that there seems to be no methods for you to manipulate those instances. You can't even do bitwise addition becuase `+` donsn't work on our new primitive type.

Reviewer: ```suggestion
But soon you will find that you don't know how to create a instance of `A` because Julia does not have a default constructor for primitive types. And if you think further, you will notice that there seems to be no methods for you to manipulate those instances. You can't even do bitwise addition becuase `+` does not work on our new primitive type.
```

> Then what can we do? Well, Julia do have functions for you to do those stuff, but most of those function are functions that generated by Julia's C++ source code or from the LLVM. So, before we started, we will take a first look at those functions lying under `Core.Intrinsics`.

Reviewer: ```suggestion
Then what can we do? Well, Julia has functions for you to do those stuff, but most of those function are generated by Julia's C++ source code or from LLVM. So, before we started, we will take a first look at those functions lying under `Core.Intrinsics`.
```

    
see [Core.Intrinsics Reference](#reference) for more functions. 

Now we have the basic function we need, we will start to define our own methds.

## Define methods

All the basic methods of our new primitve type should use those functions. For defining our own constructor, we need to construct from some built-in primitive 
type. 

For example: 

```julia
import Base: +
primitive type A 72 end
```

Here is the constructor for `A`. extend a `Int64` to 72 bit with 0 and mark as type `A`.

```julia
A(x::Int64) = Core.Intrinsics.zext_int(A, x)
```

We can also accept and truncate a `Int128` to 72 bit and mark as type `A`.

```julia
A(x::Int128) = Core.Intrinsics.trunc_int(A, x)
```

`Base.+` is overloaded for defining addition on type `A`.

```julia
+(x::A, y::A) = Core.Intrinsics.add_int(x, y)
```

Consequently, the usage of primitive type `A` could be looked like this:

```julia
a1 = A(3) #create instance of A
a2 = A(12345678901234567890)

a1 + a2 # add the value of a1 and a2
```

## Q&A

Q. What's the benefit of using a custom type? Why not just define a wrapper type of type alias of a built-in primitive type?
A. 

> First, if you want to dispatch on the type, than type alias might not work since it is actually the same type as the origin. 

Reviewer: ```suggestion
First, if you want to dispatch on the type, then a type alias will not work since it is actually the same type as the aliased one. 
```

Second, a wrapper type will cause some performance drop owing to the unwrapping process and the compiler might not be able to optimise the code. compare the result of `@code_llvm a1 + a2` on the above example and a wrapper type.

Last but not the least, You can define a custom bitsize for your primitive type and allow LLVM to do the magic.

Q. Where can I find more information?

> A. you can take a look at the [LLVM reference](https://llvm.org/docs/LangRef.html) for the LLVM Intrinsics and the [Julia c++ source code for Intrinsics](https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/src/intrinsics.cpp#L853-L1241) and [header file](https://github.com/JuliaLang/julia/blob/v1.1.0/src/intrinsics.h) for function arguments, more detail at the [codegen](https://github.com/JuliaLang/julia/blob/v1.1.0/src/codegen.cpp) and [ccall](https://github.com/JuliaLang/julia/blob/v1.1.0/src/ccall.cpp), and Julia code example in [Julia's test](https://github.com/JuliaLang/julia/blob/v1.1.0/test/intrinsics.jl) and the [defination of built-in primitive type](https://github.com/JuliaLang/julia/blob/v1.1.0/base/boot.jl).

Reviewer: ```suggestion
A. you can take a look at the [LLVM reference](https://llvm.org/docs/LangRef.html) for the LLVM Intrinsics and the [Julia C++ source code for Intrinsics](https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/src/intrinsics.cpp#L853-L1241) and [header file](https://github.com/JuliaLang/julia/blob/v1.1.0/src/intrinsics.h) for function arguments, more detail at the [codegen](https://github.com/JuliaLang/julia/blob/v1.1.0/src/codegen.cpp) and [ccall](https://github.com/JuliaLang/julia/blob/v1.1.0/src/ccall.cpp), and Julia code example in [Julia's test](https://github.com/JuliaLang/julia/blob/v1.1.0/test/intrinsics.jl) and the [defination of built-in primitive type](https://github.com/JuliaLang/julia/blob/v1.1.0/base/boot.jl).
```

## Bitboard example

> [bitboard](https://en.wikipedia.org/wiki/Bitboard) is a way of store and compute board game state in binary expression, for exampe we want to store a 4x4 board for a [2048 game](https://en.wikipedia.org/wiki/2048_(video_game)) we can do like this:

Reviewer: ```suggestion
A [bitboard](https://en.wikipedia.org/wiki/Bitboard) is a way of store and compute board game state in binary expression, for exampe we want to store a 4x4 board for a [2048 game](https://en.wikipedia.org/wiki/2048_(video_game)) we can do like this:
```

    x = xor_int(x, xor_int(t, shl_int(t, 32)))
    x
end

function reflect_horizontal(x::BitBoard128)
    t = and_int(xor_int(x, lshr_int(x, 16)), BitBoard128(0x0000ffff0000ffff0000ffff0000ffff))
    x = xor_int(x, xor_int(t, shl_int(t, 16)))
    t = and_int(xor_int(x, lshr_int(x, 8)), BitBoard128(0x00ff00ff00ff00ff00ff00ff00ff00ff))
    x = xor_int(x, xor_int(t, shl_int(t, 8)))
    x
end

function rotate_right(b::AbstractBoard)
    b = transpose(b)
    b = reflect_horizontal(b)
    b
end

function rotate_left(b::AbstractBoard)
    b = transpose(b)
    b = reflect_vertical(b)
    b
end

GetRow(t::BitBoard128, r::Int)::UInt32 = and_int(trunc_int(UInt32, lshr_int(t, 32 * (3 - r))), 0xffffffff)
GetRow(t::BitBoard64, r::Int)::UInt16 = and_int(trunc_int(UInt16, lshr_int(t, 16 * (3 - r))), 0xffff)

function empty(b::AbstractBoard)::Vector{Int}
    a = Vector{Int}()
    for i in 0:15
        if b(i) == 0
            push!(a, i)
        end
    end
    return a
end
```

<a id="reference"></a>

## Core.Intrinsics Reference

reference for some basic function under the `Core.Intrinsics`.
1.  Type construction
    
    1.  `Core.Intrinsics.bitcase(type, value)`
    

>         `bitcase` told the LLVM to make a new same sized value of a primitive type instance `value`         and mark it as type `type`. The type of `value` must have the same bitsize as `type`.

Reviewer: why these spaces between `value` and "and". Same applies to next lines

Reviewer 2: Oh, I guess it's due to some problems in my editor, let me remove it.
